third_party/move/move-compiler-v2/src/pipeline/control_flow_graph_simplifier.rs

```diff
@@ -0,0 +1,595 @@
+                this.replace_succs(pred, block_to_remove, redirect_to);
+            },
+            // replace `block_to_remove` by predecessors of `block_to_remove`
+            // in predecessors of `redirect_to`
+            |this, succ| {
+                debug_assert!(succ == redirect_to);
+                this.remove_pred(succ, block_to_remove);
+                let mut preds_of_block_to_remove = this.preds(block_to_remove).to_vec();
+                this.preds_mut(succ).append(&mut preds_of_block_to_remove);
+            },
+        );
+    }
+
+    /// Redirects a code block so that it jumps/branches to `to`
+    /// where it originally jumps/branches to `from`.
+    /// Does nothing if `code` doesn't end with a jump/branch
+    /// Requries: `code` not empty
+    fn redirects_block(code: &mut [Bytecode], from: Label, to: Label) {
+        let last_instr = code.last_mut().expect("last instruction");
+        match last_instr {
+            Bytecode::Branch(_, l0, l1, _) => {
+                subst_label(l0, from, to);
+                subst_label(l1, from, to);
+            },
+            Bytecode::Jump(_, label) => {
+                subst_label(label, from, to);
+            },
+            _ => {},
+        }
+    }
+}
+
+/// Transformation state for transformation 2 in the module doc
+struct RedundantJumpRemover(pub ControlFlowGraphCodeGenerator);
+
+impl RedundantJumpRemover {
+    /// Wrapper
+    pub fn new(cfg_generator: ControlFlowGraphCodeGenerator) -> Self {
+        Self(cfg_generator)
+    }
+
+    /// Performs the transformation 2 described in the module doc
+    pub fn transform(&mut self) {
+        for block in self.0.blocks() {
+            // the later condition says that `block` is unreachable or has been removed
+            if self.0.is_trivial_block(block) || !self.0.predecessors.contains_key(&block) {
+                continue;
+            }
+            self.transform_edges_from(block)
+        }
+    }
+
+    /// Removes all redundant edges from `block`
+    /// Requires: `block` still in the cfg, `block` is not entry/exit
+    fn transform_edges_from(&mut self, block: BlockId) {
+        for suc in self.0.successors.get(&block).expect("successors").clone() {
+            if !self.0.is_trivial_block(suc) && self.remove_jump_if_possible(block, suc) {
+                // successors of `block` changes
+                // consider L0: goto L1; L1: goto L2; L2: goto L3;
+                // suc L0: L1
+                // after merging block L1 into block L0
+                // L0: goto L2; L2: goto L3;
+                // suc L0: L2
+                // we continue to merge block L3 into block L0
+                self.transform_edges_from(block);
+            }
+        }
+    }
+
+    /// An edge can be removed if `from` has only one successor, and `to` has only one predecessor
+    fn can_remove_edge(&self, from: BlockId, to: BlockId) -> bool {
```

Comment:
It seems that this function removes the edge "from -> to" when `from` has only `to` as the successor and `to` has only `from` as the predecessor. If it is the case, maybe rename the function with a more specific name to indicate the purpose and add the check `self.0.succs(from).contains(&to)) && self.0.preds(&to).contains(from)`? 
Also, is adding `debug_assert!` a common practice to enforce the invariant of a CFG in general? Just curious.

Reply:
Added check for `self.0.succs(from).contains(&to))`. The latter is an invariant of the CFG.

I don't know if there is a term for an edge where the source has only one successor and the destination has only one predecessor. This sounds like the opposite of a critical edge though.

Wolfgang was using debug_assert! to check some invariants of the borrow graph in the old borrow checker. Not sure if this practice is followed anywhere else. But I do think this module is tricky enough that we should add such checks.